File: NPPContractManagement.API/Services/MemberAccountService.cs
using NPPContractManagement.API.DTOs;
using NPPContractManagement.API.Models;
using NPPContractManagement.API.Repositories;

namespace NPPContractManagement.API.Services
{
    public class MemberAccountService : IMemberAccountService
    {
        private readonly IMemberAccountRepository _memberAccountRepository;
        private readonly IIndustryRepository _industryRepository;

        public MemberAccountService(IMemberAccountRepository memberAccountRepository, IIndustryRepository industryRepository)
        {
            _memberAccountRepository = memberAccountRepository;
            _industryRepository = industryRepository;
        }

        public async Task<IEnumerable<MemberAccountDto>> GetAllMemberAccountsAsync()
        {
            var memberAccounts = await _memberAccountRepository.GetAllAsync();
            return memberAccounts.Select(MapToDto);
        }

        public async Task<MemberAccountDto?> GetMemberAccountByIdAsync(int id)
        {
            var memberAccount = await _memberAccountRepository.GetByIdAsync(id);
            return memberAccount != null ? MapToDto(memberAccount) : null;
        }

        public async Task<MemberAccountDto> CreateMemberAccountAsync(CreateMemberAccountDto createMemberAccountDto, string createdBy)
        {
            // Validate unique member number
            var exists = await _memberAccountRepository.ExistsByMemberNumberAsync(createMemberAccountDto.MemberNumber);
            if (exists)
            {
                throw new ArgumentException("Member number already exists", nameof(createMemberAccountDto.MemberNumber));
            }

            // Validate industry exists if provided
            if (createMemberAccountDto.IndustryId.HasValue)
            {
                var industry = await _industryRepository.GetByIdAsync(createMemberAccountDto.IndustryId.Value);
                if (industry == null)
                {
                    throw new ArgumentException("Industry not found", nameof(createMemberAccountDto.IndustryId));
                }
            }

            var memberAccount = new MemberAccount
            {
                MemberNumber = createMemberAccountDto.MemberNumber,
                FacilityName = createMemberAccountDto.FacilityName,
                Address = createMemberAccountDto.Address,
                City = createMemberAccountDto.City,
                State = createMemberAccountDto.State,
                ZipCode = createMemberAccountDto.ZipCode,
                Country = createMemberAccountDto.Country,
                PhoneNumber = createMemberAccountDto.PhoneNumber,

                IndustryId = createMemberAccountDto.IndustryId,
                W9 = createMemberAccountDto.W9,
                W9Date = createMemberAccountDto.W9Date,
                TaxId = createMemberAccountDto.TaxId,
                BusinessType = createMemberAccountDto.BusinessType,
                // New fields
                LopDate = createMemberAccountDto.LopDate ?? DateTime.UtcNow.Date,
                InternalNotes = createMemberAccountDto.InternalNotes,
                ClientGroupEnrollment = createMemberAccountDto.ClientGroupEnrollment,
                SalesforceAccountName = string.IsNullOrWhiteSpace(createMemberAccountDto.SalesforceAccountName)
                    ? createMemberAccountDto.FacilityName
                    : createMemberAccountDto.SalesforceAccountName,
                VMAPNumber = createMemberAccountDto.VMAPNumber,
                VMSupplierName = createMemberAccountDto.VMSupplierName,
                VMSupplierSite = createMemberAccountDto.VMSupplierSite,
                PayType = createMemberAccountDto.PayType,
                ParentMemberAccountNumber = createMemberAccountDto.ParentMemberAccountNumber,
                EntegraGPONumber = createMemberAccountDto.EntegraGPONumber,
                ClientGroupNumber = createMemberAccountDto.ClientGroupNumber,
                EntegraIdNumber = createMemberAccountDto.EntegraIdNumber,
                AuditDate = DateTime.UtcNow,
                Status = (MemberAccountStatus)createMemberAccountDto.Status,
                IsActive = true,
                CreatedDate = DateTime.UtcNow,
                CreatedBy = createdBy
            };

            var createdMemberAccount = await _memberAccountRepository.AddAsync(memberAccount);
            return MapToDto(createdMemberAccount);
        }

        public async Task<MemberAccountDto> UpdateMemberAccountAsync(int id, UpdateMemberAccountDto updateMemberAccountDto, string modifiedBy)
        {
            var memberAccount = await _memberAccountRepository.GetByIdAsync(id);
            if (memberAccount == null)
            {
                throw new ArgumentException("Member account not found", nameof(id));
            }

            // Validate unique member number
            var exists = await _memberAccountRepository.ExistsByMemberNumberAsync(updateMemberAccountDto.MemberNumber, id);
            if (exists)
            {
                throw new ArgumentException("Member number already exists", nameof(updateMemberAccountDto.MemberNumber));
            }

            // Validate industry exists if provided
            if (updateMemberAccountDto.IndustryId.HasValue)
            {
                var industry = await _industryRepository.GetByIdAsync(updateMemberAccountDto.IndustryId.Value);
                if (industry == null)
                {
                    throw new ArgumentException("Industry not found", nameof(updateMemberAccountDto.IndustryId));
                }
            }

            memberAccount.MemberNumber = updateMemberAccountDto.MemberNumber;
            memberAccount.FacilityName = updateMemberAccountDto.FacilityName;
            memberAccount.Address = updateMemberAccountDto.Address;
            memberAccount.City = updateMemberAccountDto.City;
            memberAccount.State = updateMemberAccountDto.State;
            memberAccount.ZipCode = updateMemberAccountDto.ZipCode;
            memberAccount.Country = updateMemberAccountDto.Country;
            memberAccount.PhoneNumber = updateMemberAccountDto.PhoneNumber;

            memberAccount.IndustryId = updateMemberAccountDto.IndustryId;
            memberAccount.W9 = updateMemberAccountDto.W9;
            memberAccount.W9Date = updateMemberAccountDto.W9Date;
            memberAccount.TaxId = updateMemberAccountDto.TaxId;
            memberAccount.BusinessType = updateMemberAccountDto.BusinessType;
            // New fields
            memberAccount.LopDate = updateMemberAccountDto.LopDate ?? memberAccount.LopDate ?? DateTime.UtcNow.Date;
            memberAccount.InternalNotes = updateMemberAccountDto.InternalNotes;
            memberAccount.ClientGroupEnrollment = updateMemberAccountDto.ClientGroupEnrollment;
            memberAccount.SalesforceAccountName = string.IsNullOrWhiteSpace(updateMemberAccountDto.SalesforceAccountName)
                ? memberAccount.FacilityName
                : updateMemberAccountDto.SalesforceAccountName;
            memberAccount.VMAPNumber = updateMemberAccountDto.VMAPNumber;
            memberAccount.VMSupplierName = updateMemberAccountDto.VMSupplierName;
            memberAccount.VMSupplierSite = updateMemberAccountDto.VMSupplierSite;
            memberAccount.PayType = updateMemberAccountDto.PayType ?? memberAccount.PayType;
            memberAccount.ParentMemberAccountNumber = updateMemberAccountDto.ParentMemberAccountNumber;
            memberAccount.EntegraGPONumber = updateMemberAccountDto.EntegraGPONumber;
            memberAccount.ClientGroupNumber = updateMemberAccountDto.ClientGroupNumber;
            memberAccount.EntegraIdNumber = updateMemberAccountDto.EntegraIdNumber;
            memberAccount.AuditDate = DateTime.UtcNow;
            memberAccount.Status = (MemberAccountStatus)updateMemberAccountDto.Status;
            memberAccount.IsActive = updateMemberAccountDto.IsActive;
            memberAccount.ModifiedDate = DateTime.UtcNow;
            memberAccount.ModifiedBy = modifiedBy;

            var updatedMemberAccount = await _memberAccountRepository.UpdateAsync(memberAccount);
            return MapToDto(updatedMemberAccount);
        }

        public async Task<bool> DeleteMemberAccountAsync(int id)
        {
            var memberAccount = await _memberAccountRepository.GetByIdAsync(id);
            if (memberAccount == null)
            {
                return false;
            }

            // Check if member account has customer accounts
            if (memberAccount.CustomerAccounts?.Any() == true)
            {
                throw new InvalidOperationException("Cannot delete member account with existing customer accounts");
            }

            await _memberAccountRepository.DeleteAsync(id);
            return true;
        }

        public async Task<bool> ActivateMemberAccountAsync(int id, string modifiedBy)
        {
            var memberAccount = await _memberAccountRepository.GetByIdAsync(id);
            if (memberAccount == null)
            {
                return false;
            }

            memberAccount.IsActive = true;
            memberAccount.Status = MemberAccountStatus.Active;
            memberAccount.ModifiedDate = DateTime.UtcNow;
            memberAccount.ModifiedBy = modifiedBy;

            await _memberAccountRepository.UpdateAsync(memberAccount);
            return true;
        }

        public async Task<bool> DeactivateMemberAccountAsync(int id, string modifiedBy)
        {
            var memberAccount = await _memberAccountRepository.GetByIdAsync(id);
            if (memberAccount == null)
            {
                return false;
            }

            memberAccount.IsActive = false;
            //memberAccount.Status = MemberAccountStatus.Inactive;
            memberAccount.ModifiedDate = DateTime.UtcNow;
            memberAccount.ModifiedBy = modifiedBy;

            await _memberAccountRepository.UpdateAsync(memberAccount);
            return true;
        }

        public async Task<MemberAccountDto?> GetMemberAccountByMemberNumberAsync(string memberNumber)
        {
            var memberAccount = await _memberAccountRepository.GetByMemberNumberAsync(memberNumber);
            return memberAccount != null ? MapToDto(memberAccount) : null;
        }

        public async Task<IEnumerable<MemberAccountDto>> GetMemberAccountsByIndustryIdAsync(int industryId)
        {
            var memberAccounts = await _memberAccountRepository.GetByIndustryIdAsync(industryId);
            return memberAccounts.Select(MapToDto);
        }

        public async Task<IEnumerable<MemberAccountDto>> GetMemberAccountsByStatusAsync(int status)
        {
            var memberAccounts = await _memberAccountRepository.GetByStatusAsync((MemberAccountStatus)status);
            return memberAccounts.Select(MapToDto);
        }

        public async Task<(IEnumerable<MemberAccountDto> MemberAccounts, int TotalCount)> SearchMemberAccountsAsync(string searchTerm, int? industryId = null, int? status = null, string? w9 = null, string? state = null, int page = 1, int pageSize = 10)
        {
            var memberAccounts = await _memberAccountRepository.SearchAsync(searchTerm, industryId, status.HasValue ? (MemberAccountStatus)status.Value : null, w9, state, page, pageSize);
            var totalCount = await _memberAccountRepository.GetCountAsync(searchTerm, industryId, status.HasValue ? (MemberAccountStatus)status.Value : null, w9, state);

            return (memberAccounts.Select(MapToDto), totalCount);
        }

        private static MemberAccountDto MapToDto(MemberAccount memberAccount)
        {
            return new MemberAccountDto
            {
                Id = memberAccount.Id,
                MemberNumber = memberAccount.MemberNumber,
                FacilityName = memberAccount.FacilityName,
                Address = memberAccount.Address,
                City = memberAccount.City,
                State = memberAccount.State,
                ZipCode = memberAccount.ZipCode,
                Country = memberAccount.Country,
                PhoneNumber = memberAccount.PhoneNumber,

                IndustryId = memberAccount.IndustryId,
                IndustryName = memberAccount.Industry?.Name,
                W9 = memberAccount.W9,
                W9Date = memberAccount.W9Date,
                TaxId = memberAccount.TaxId,
                BusinessType = memberAccount.BusinessType,
                // New fields
                LopDate = memberAccount.LopDate,
                InternalNotes = memberAccount.InternalNotes,
                ClientGroupEnrollment = memberAccount.ClientGroupEnrollment,
                SalesforceAccountName = memberAccount.SalesforceAccountName,
                VMAPNumber = memberAccount.VMAPNumber,
                VMSupplierName = memberAccount.VMSupplierName,
                VMSupplierSite = memberAccount.VMSupplierSite,
                PayType = memberAccount.PayType,
                PayTypeName = memberAccount.PayType,
                ParentMemberAccountNumber = memberAccount.ParentMemberAccountNumber,
                EntegraGPONumber = memberAccount.EntegraGPONumber,
                ClientGroupNumber = memberAccount.ClientGroupNumber,
                EntegraIdNumber = memberAccount.EntegraIdNumber,
                AuditDate = memberAccount.AuditDate,
                Status = (int)memberAccount.Status,
                StatusName = memberAccount.Status.ToString(),
                IsActive = memberAccount.IsActive,
                CreatedDate = memberAccount.CreatedDate,
                ModifiedDate = memberAccount.ModifiedDate,
                CreatedBy = memberAccount.CreatedBy,
                ModifiedBy = memberAccount.ModifiedBy,
                CustomerAccountsCount = memberAccount.CustomerAccounts?.Count ?? 0
            };
        }
    }
}

File: NPPContractManagement.Frontend/src/app/services/member-account.service.ts
import { Injectable } from '@angular/core';
import { Observable, throwError } from 'rxjs';
import { catchError, map } from 'rxjs/operators';
import { ApiService } from './api.service';
import {
  MemberAccount,
  CreateMemberAccountRequest,
  UpdateMemberAccountRequest,
  MemberAccountSearchResult,
  MemberAccountStatus
} from '../models/member-account.model';

export interface PaginatedResult<T> {
  items: T[];
  totalCount: number;
  pageNumber: number;
  pageSize: number;
  totalPages: number;
}

@Injectable({
  providedIn: 'root'
})
export class MemberAccountService {
  private readonly endpoint = 'member-accounts';

  constructor(private apiService: ApiService) {}

  getAllMemberAccounts(): Observable<MemberAccount[]> {
    return this.apiService.get<MemberAccount[]>(this.endpoint);
  }

  getMemberAccountById(id: number): Observable<MemberAccount> {
    return this.apiService.get<MemberAccount>(`${this.endpoint}/${id}`);
  }

  createMemberAccount(memberAccount: CreateMemberAccountRequest): Observable<MemberAccount> {
    return this.apiService.post<MemberAccount>(this.endpoint, memberAccount);
  }

  updateMemberAccount(id: number, memberAccount: UpdateMemberAccountRequest): Observable<MemberAccount> {
    return this.apiService.put<MemberAccount>(`${this.endpoint}/${id}`, memberAccount);
  }

  deleteMemberAccount(id: number): Observable<any> {
    return this.apiService.delete(`${this.endpoint}/${id}`);
  }

  activateMemberAccount(id: number): Observable<any> {
    return this.apiService.post(`${this.endpoint}/${id}/activate`, {});
  }

  suspendMemberAccount(id: number): Observable<any> {
    return this.apiService.post(`${this.endpoint}/${id}/suspend`, {});
  }

  updateW9Status(id: number, w9: boolean, w9Date?: Date): Observable<any> {
    return this.apiService.post(`${this.endpoint}/${id}/w9-status`, { w9, w9Date });
  }

  searchMemberAccounts(params: any): Observable<MemberAccountSearchResult> {
    return this.apiService.get<MemberAccountSearchResult>(`${this.endpoint}/search`, params);
  }

  getMemberAccountsByIndustry(industryId: number): Observable<MemberAccount[]> {
    return this.apiService.get<MemberAccount[]>(`${this.endpoint}/by-industry/${industryId}`);
  }

  getMemberAccountsWithoutW9(): Observable<MemberAccount[]> {
    return this.apiService.get<MemberAccount[]>(`${this.endpoint}/without-w9`);
  }

  exportToExcel(params?: any): Observable<Blob> {
    return this.apiService.getBlob(`${this.endpoint}/export`, params);
  }

  bulkDelete(ids: number[]): Observable<any> {
    return this.apiService.post(`${this.endpoint}/bulk-delete`, { ids });
  }

  bulkUpdateStatus(ids: number[], status: string): Observable<any> {
    return this.apiService.post(`${this.endpoint}/bulk-update-status`, { ids, status });
  }

  // Alias methods for compatibility
  getAll(): Observable<MemberAccount[]> {
    return this.getAllMemberAccounts();
  }

  getById(id: number): Observable<MemberAccount> {
    return this.getMemberAccountById(id);
  }

  create(memberAccount: CreateMemberAccountRequest): Observable<MemberAccount> {
    return this.createMemberAccount(memberAccount);
  }

  getPaginated(
    pageNumber: number,
    pageSize: number,
    sortBy?: string,
    sortDirection: 'asc' | 'desc' = 'asc',
    searchTerm?: string,
    status?: string,
    industryId?: number,
    state?: string
  ): Observable<PaginatedResult<MemberAccount>> {
    const params: any = {
      pageNumber,
      pageSize,
      sortDirection
    };
    if (sortBy) params.sortBy = sortBy;
    if (searchTerm) params.searchTerm = searchTerm;
    if (status) params.status = status;
    if (industryId) params.industryId = industryId;
    if (state) params.state = state;

    return this.apiService.get<any>(this.endpoint, params).pipe(
      map(response => {
        const items = (response.Items || response.items || []).map((dto: any) => this.mapDtoToMemberAccount(dto));
        return {
          items,
          totalCount: response.TotalCount || response.totalCount || items.length,
          pageNumber: response.PageNumber || response.pageNumber || pageNumber,
          pageSize: response.PageSize || response.pageSize || pageSize,
          totalPages: response.TotalPages || response.totalPages || Math.ceil((response.TotalCount || items.length) / (response.PageSize || pageSize))
        } as PaginatedResult<MemberAccount>;
      }),
      catchError(err => throwError(() => err))
    );
  }

  private mapDtoToMemberAccount(dto: any): MemberAccount {
    // Map status
    let status: MemberAccountStatus = MemberAccountStatus.Active;
    if (dto.StatusName || dto.statusName) {
      const name = dto.StatusName || dto.statusName;
      switch (name) {
        case 'Active': status = MemberAccountStatus.Active; break;
        case 'Inactive': status = MemberAccountStatus.Inactive; break;
        case 'Pending': status = MemberAccountStatus.Pending; break;
        case 'Suspended': status = MemberAccountStatus.Suspended; break;
        default: status = MemberAccountStatus.Active;
      }
    } else {
      const val = dto.Status || dto.status || 1;
      switch (val) {
        case 1: status = MemberAccountStatus.Active; break;
        case 2: status = MemberAccountStatus.Inactive; break;
        case 3: status = MemberAccountStatus.Pending; break;
        case 4: status = MemberAccountStatus.Suspended; break;
        default: status = MemberAccountStatus.Active;
      }
    }

    return {
      id: dto.Id || dto.id,
      memberNumber: dto.MemberNumber || dto.memberNumber,
      facilityName: dto.FacilityName || dto.facilityName,
      address: dto.Address || dto.address,
      city: dto.City || dto.city,
      state: dto.State || dto.state,
      zipCode: dto.ZipCode || dto.zipCode,
      country: dto.Country || dto.country,
      phoneNumber: dto.PhoneNumber || dto.phoneNumber,

      industryId: dto.IndustryId || dto.industryId,
      industryName: dto.IndustryName || dto.industryName,
      w9: dto.W9 ?? dto.w9 ?? false,
      w9Date: dto.W9Date || dto.w9Date ? new Date(dto.W9Date || dto.w9Date) : undefined,
      taxId: dto.TaxId || dto.taxId,
      businessType: dto.BusinessType || dto.businessType,
      status: status,
      isActive: dto.IsActive !== undefined ? dto.IsActive : (dto.isActive ?? true),
      createdDate: new Date(dto.CreatedDate || dto.createdDate || Date.now()),
      modifiedDate: dto.ModifiedDate || dto.modifiedDate ? new Date(dto.ModifiedDate || dto.modifiedDate) : undefined,
      createdBy: dto.CreatedBy || dto.createdBy,
      modifiedBy: dto.ModifiedBy || dto.modifiedBy,
      // Extended fields
      lopDate: dto.LopDate || dto.lopDate ? new Date(dto.LopDate || dto.lopDate) : undefined,
      internalNotes: dto.InternalNotes || dto.internalNotes,
      clientGroupEnrollment: (() => { const v = (dto.ClientGroupEnrollment ?? dto.clientGroupEnrollment); return (typeof v === 'number') ? v : (v === true ? 1 : (v === false ? 0 : undefined)); })(),
      salesforceAccountName: dto.SalesforceAccountName || dto.salesforceAccountName,
      vmapNumber: dto.VMAPNumber || dto.vmapNumber,
      vmSupplierName: dto.VMSupplierName || dto.vmSupplierName,
      vmSupplierSite: dto.VMSupplierSite || dto.vmSupplierSite,
      payType: dto.PayType || dto.payType,
      payTypeName: dto.PayTypeName || dto.payTypeName,
      parentMemberAccountNumber: dto.ParentMemberAccountNumber || dto.parentMemberAccountNumber,
      entegraGPONumber: dto.EntegraGPONumber || dto.entegraGPONumber,
      clientGroupNumber: dto.ClientGroupNumber || dto.clientGroupNumber,
      entegraIdNumber: dto.EntegraIdNumber || dto.entegraIdNumber,
      auditDate: dto.AuditDate || dto.auditDate ? new Date(dto.AuditDate || dto.auditDate) : undefined,
      customerAccountsCount: dto.CustomerAccountsCount || dto.customerAccountsCount
    } as MemberAccount;
  }


  update(id: number, memberAccount: UpdateMemberAccountRequest): Observable<MemberAccount> {
    return this.updateMemberAccount(id, memberAccount);
  }

  delete(id: number): Observable<any> {
    return this.deleteMemberAccount(id);
  }
}
